gunrock/util/kernel_segmode.hxx

```diff
@@ -0,0 +1,1259 @@
+      }
+        
+      else if(b.left_element < b.best_count){
+      a.best_element = a.best_count >= b.best_count?a.best_element:b.best_element;
+      a.best_count = max(a.best_count, b.best_count);
+      }
+
+      // else if(a.current_count <= a.best_count){
+      // we don't need to change anything}
+
+      if (b.right_element == a.right_element){
+        b.right_count += a.right_count;
+      }
+    }
+
+    // they don't overlap
+    // which means we need to compare the two bests for the bests
+    else if (b.left_element != a.right_element){
+      a.best_element = a.best_count >= b.best_count?a.best_element:b.best_element;
+      a.best_count = max(a.best_count, b.best_count);
+    }
+
+    // update the left
+    if (a.left_element == a.best_element){
+      a.left_count = a.best_count;
+    }
+
+    // update the right
+    // if b.right_element == b.left_element or even b.best_element
+    a.right_element = b.right_element;
+
+    if (a.right_element == a.best_element){
+      a.right_count = a.best_count;
+    }
+    else{
+      a.right_count = b.right_count;
+    }
+    
+    return a;
+  }
+  else{
+       if (a.left_element == b.right_element){
+      
+      a.left_element = b.right_count + a.left_count;
+      if (a.left_element > b.best_count && a.left_element >= a.best_count){
+        b.best_element = b.right_element;
+        b.best_count = a.left_element;
+      }
+        
+      else if(a.left_element < a.best_count){
+        b.best_count = max(b.best_count, a.best_count);
+      b.best_element = b.best_count >= a.best_count?b.best_element:a.best_element;
+      }
+   
+      if (a.right_element == b.right_element){
+          a.right_count += b.right_count;
+        }
+    
+    }
+   else{
+      b.best_count = max(b.best_count, a.best_count);
+      b.best_element = b.best_count >= a.best_count?b.best_element:a.best_element;
+    }
+
+    if (b.left_element == b.best_element){
+      b.left_count = b.best_count;
+    }
+    b.right_element = a.right_element;
+    if (b.right_element == b.best_element){
+      b.right_count =b.best_count;
+    }
+    else{
+      b.right_count = a.right_count;
+    }
+    
+    return b;
+  }
+  }
+
+
+  MGPU_HOST_DEVICE quad operator()(quad a, int b) const {
+
+    if (b == a.left_element){
+    
+      a.left_count++;
+      a.right_count++;
+      if(b == a.best_element) a.best_count++;
+    }
+
+    else if( b == a.right_element){
+      a.right_count++;
+      
+    }
+    else if( b != a.right_element){
+      a.best_element = a.right_count > a.best_count? a.right_element:a.best_element;
+      a.best_count = max(a.right_count, a.best_count);
+
+      a.right_count = 1;
+      a.right_element = b;
+    }
+
+  
+    return a;
+  }
+
+};
+
+
+
+
+
+
+template<mgpu::bounds_t bounds, typename a_keys_it, typename b_keys_it,
+  typename comp_t>
+mgpu::mem_t<int> gunrock::util::merge_path_partitions(a_keys_it a, int64_t a_count, b_keys_it b,
```

Comment:
The namespaces here are confusing. Could you be more clear about where these functions should be defined?